Development/nmos/constraints.cpp

```diff
@@ -0,0 +1,165 @@
+                return false;
+            }
+            if (constraint.has_field(nmos::fields::constraint_minimum) && subconstraint.has_field(nmos::fields::constraint_minimum))
+            {
+                if (constraint.at(U("minimum")).has_field(nmos::fields::numerator))
+                {
+                    if (nmos::parse_rational(constraint.at(U("minimum"))) > nmos::parse_rational(subconstraint.at(U("minimum"))))
+                    {
+                        return false;
+                    }
+                }
+                else if (nmos::fields::constraint_minimum(constraint) > nmos::fields::constraint_minimum(subconstraint))
+                {
+                    return false;
+                }
+            }
+            if (constraint.has_field(nmos::fields::constraint_maximum) && subconstraint.has_field(nmos::fields::constraint_maximum))
+            {
+                if (constraint.at(U("maximum")).has_field(nmos::fields::numerator))
+                {
+                    if (nmos::parse_rational(constraint.at(U("maximum"))) < nmos::parse_rational(subconstraint.at(U("maximum"))))
+                    {
+                        return false;
+                    }
+                }
+                else if (nmos::fields::constraint_maximum(constraint) < nmos::fields::constraint_maximum(subconstraint))
+                {
+                    return false;
+                }
+            }
+            // subconstraint enum values should match constraint
+            if (subconstraint.has_field(nmos::fields::constraint_enum))
+            {
+                const auto& subconstraint_enum_values = nmos::fields::constraint_enum(subconstraint).as_array();
+                if (subconstraint_enum_values.end() == std::find_if(subconstraint_enum_values.begin(), subconstraint_enum_values.end(), [&constraint](const web::json::value& enum_value)
+                {
+                    return match_constraint(enum_value, constraint);
+                }))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        // Constraint Set B is a subset of Constraint Set A if all of Parameter Constraints of Constraint Set B, except for meta, are present in Constraint Set A and each Parameter Constraint of Constraint Set B, except for meta, is a subconstraint of the according Parameter Constraint of Constraint Set A.
+        // Constraint B is a subconstraint of a Constraint A if:
+
+        // 1. Constraint B has enum keyword when Constraint A has it and enum of Constraint B is a subset of enum of Constraint A
+        // 2. Constraint B has enum or minimum keyword when Constraint A has minimum keyword and allowed values for Constraint B are less than allowed values for Constraint A
+        // 3. Constraint B has enum or maximum keyword when Constraint A has maximum keyword and allowed values for Constraint B are greater than allowed values for Constraint A
```

Comment:
"Constraint B has enum or minimum keyword when Constraint A has minimum keyword and allowed values for Constraint B are less than allowed values for Constraint A"

I'm not sure this (and the related description of maximum) is right either... Constraint B can only be a subconstraint (i.e. be more constrained) if its minimum is _greater_ than the minimum permitted by Constraint A, right?

```suggestion
        // Constraint B is a subconstraint of a Constraint A if:
        // 1. Constraint B has enum keyword when Constraint A has it and enumerated values of Constraint B are a subset of enumerated values of Constraint A
        // 2. Constraint B has enum or minimum keyword when Constraint A has minimum keyword and allowed values of Constraint B are greater than minimum value of Constraint A
        // 3. Constraint B has enum or maximum keyword when Constraint A has maximum keyword and allowed values of Constraint B are less than maximum value of Constraint A
```

Reply:
Fixed the description.